Ticket Manager Client + Server

ZBejavu opened the pull request:
added features : 
- view more or less options for longer pieces of ticket description
- sorting the tickets by the creation time - both ways.
- filtering the tickets to a specific date range (easier to handle and gives room for more specific handling).
- mailing a response to the mail of the ticket creator by clicking on the email link inside the ticket.
- handling a ticket by closing it, leaving room for inputting additional information, reason for closure and requiring a name/id to correspond with the handling of the ticket (so that Wix can keep track of what their employees do)

- the handled tickets can be seen by pressing the corresponding button at the top of the ticket list, there is an option to reopen the ticket.

- iv'e added a test that checks functionality of the ticket handling in the file by name of 'myTest.test.js'.

** the eslint really does a lot for organizing and readability.

ZBejavu committed:
initial installation

ZBejavu committed:
hopefully last eslint based push commit

ZBejavu committed:
server tests work except for the last one

to be continued..

ZBejavu committed:
added styling and 'ShowMore' button

ZBejavu committed:
first 3 tests pass

ZBejavu committed:
client test pass, added translated time

ZBejavu committed:
fixed filter content bug

the wrong descritipion would show

ZBejavu committed:
corrected the time

ZBejavu committed:
fixed test - to be reverted

ZBejavu committed:
Revert "fixed test - to be reverted"

This reverts commit 7eb2e2a71e87fad23304db710f3720e2572e20b8.

ZBejavu committed:
fixing test 3

after getting the approval... used readFileSync to syncronisly read the data  as the test goes.

ZBejavu committed:
made content appear as the ticket sender intended intended

made the line breaks in the content string work.

ZBejavu committed:
Sort by Date and Filter by Date work

all test pass.
user can choose to see only Tickets from a certain date to a certain date.
this helps narrow down the searches.
you can also sort the list from newest to oldest and the other way around.

ZBejavu committed:
switched sort to option list

made the sort criteria stay after actions

ZBejavu committed:
filter + dateRange are compatible

ZBejavu committed:
added Closing Tickets

 closing the ticket by inputing an employer Id, reason for closing, and room for additional information. it is then sent and saved in the server.

the handeled tickets can be displayed buy the corresponding button on screen. there is also an option to reopen a ticket. the reopened ticket will be more visible to the user since it must have some sort of importance for it to be reopened.

ZBejavu committed:
almost finished testing the done feature

ZBejavu committed:
checking to see if tests pass in PR

ZBejavu committed:
test for handling Ticket Process works well

checking if the modal openes on click. then inputing the information. submiting, checking to see the list counter has changed, switching to the handled list and checking that the information inside the fields is identical to the information i filled out. , then opening it up again and sending the information back(reopening)

ZBejavu committed:
using the sort in the test

ZBejavu committed:
switched test to a mocking version so it will work in github

ZBejavu committed:
reverting changed data info

ZBejavu committed:
fixed bug

ZBejavu committed:
maybe headless is the issue

ZBejavu committed:
added gif

ZBejavu committed:
added my gif and Glitch Link

ZBejavu committed:
improved performance

tried to minimize the amount of times going over the ticket array. im sure there are still ways to improve but this is a lot better than before.

ZBejavu committed:
changed the reveal setting aswell

ZBejavu committed:
fixed issue when handling ticket

tickets would return to defalut after closing (getting data from server) changed it so that the done property will be updated locallly aswell so the search parameters and hidden status's would stay the same

Cyber4sPopo commented on server/app.js:
```diff
@@ -1,5 +1,87 @@
+    }
+  });
+  if (!found) {
+    res.status(404).send('not found');
```
Very nice. But you do not handle this error in the client side

Cyber4sPopo commented on server/app.js:
```diff
@@ -1,5 +1,87 @@
+        tickets[i].additional = req.query.additional;
+        delete tickets[i].reason;
+        const dataInJson = JSON.stringify(tickets, null, 2);
+        fs.writeFile(filePath, dataInJson, (e) => {
```
`fs.writeFileSync` is better here

Cyber4sPopo commented on server/app.js:
```diff
@@ -1,5 +1,87 @@
+        delete tickets[i].done;
+        tickets[i].employe = req.query.employe;
+        tickets[i].additional = req.query.additional;
+        delete tickets[i].reason;
```
why delete and not set to `false`?

ZBejavu replied:
If you set it to false, the test you designed won't work on the first try. 
It will compare the initial value of the done property (which is undefined) to the value after undone (which is false).
After the first try the test will pass since it goes over the same ticket (value will start with 'false' and end with 'false')

Cyber4sPopo commented on server/app.js:
```diff
@@ -1,5 +1,87 @@
+      const prevState = tickets[i].done;
+      found = true;
+      // console.log(tickets[i]);
+      if (!prevState) {
```
Why not update nonetheless?

ZBejavu replied:
since I'm writing the whole file from scratch  each 'writefile' command, I would prefer to write only when I actually change the data

Cyber4sPopo commented on server/app.js:
```diff
@@ -1,5 +1,87 @@
+  const data = fs.readFileSync(filePath);
+  const tickets = JSON.parse(data);
+  let found = false;
+  tickets.forEach(async (ticket, i) => {
```
Other Array functions might be better in this scenario. try `findIndex` or `map`

Cyber4sPopo commented on server/app.js:
```diff
@@ -1,5 +1,87 @@
+  }
+});
+
+app.post('/api/tickets/:ticketId/done', (req, res) => {
```
Join paths to a single path with a url param `/api/tickets/:ticketId/:done` instead of using two paths

ZBejavu committed:
fixed ticket hiding functionality

the counter is unique to each list, when you restore hidden tickets they restore only in the designated list. and performance is better

ZBejavu committed:
changed content to be less complicated

splitting it to array only in the show more..

Cyber4sPopo commented on client/src/App.js:
```diff
@@ -1,12 +1,230 @@
+  const [showFinished, setShowFinished] = useState(false);
+  // const ActiveTickets = sortByTime[0].filter((ticket) => !ticket.done && !ticket.hidden);
+  // const HandeledTickets = sortByTime[0].filter((ticket) => ticket.done && !ticket.hidden);
+  let ActiveTickets =[], HandeledTickets = [],hiddenCounter1=0 , hiddenCounter2 =0;
```
variable names should be camelCase

Cyber4sPopo commented on client/src/App.js:
```diff
@@ -1,12 +1,230 @@
+  function oldOrNew(val, arr) {
+    const arrCopy = arr ? arr.slice() : TicketArr.slice();
+    const newArr = arrCopy.filter((ticket) => ticket.creationTime < dateRange.to && ticket.creationTime > dateRange.from);
+    if (val === '0') {
```
What is this `val` variable? I cant understand from the name or from the value.
Why is its value is a stringed number? what does it represents? cant it be represented better in other format?

Cyber4sPopo commented on client/src/App.js:
```diff
@@ -1,12 +1,230 @@
+  })
+  //const hiddenCounter1 = sortByTime[0].length - (ActiveTickets.length + HandeledTickets.length);
+
+  function oldOrNew(val, arr) {
```
bad function name

Cyber4sPopo commented on client/src/App.js:
```diff
@@ -1,12 +1,230 @@
+    });
+  }
+
+    function finishLocaly(id, bool){
```
Indentation is weird

Cyber4sPopo commented on client/src/App.js:
```diff
@@ -1,12 +1,230 @@
+  }
+
+  useEffect(() => {
+    axios.get('/api/tickets').then((response) => { setSortByTime([response.data, 'default']); setTicketArr(response.data); });
```
Open a proper block if you have more then one statement

Cyber4sPopo commented on client/src/App.js:
```diff
@@ -1,12 +1,230 @@
+    newArr.forEach((ticket, i) => {
+      if (ticket.id === id) {
+        newArr[i].hidden = true;
+        setSortByTime([newArr, sortByTime[1]]);
```
Very bad pattern. you are trying to find the correct ticket, `forEach` is not your best option. and even if it was, you should do something with that ticket when the loop ends and not in the middle of it.

* look for other places in your code you used this pattern and fix them
* 👨🏻‍🏫 - prepare 5 min explanation

Cyber4sPopo commented on client/src/App.js:
```diff
@@ -1,12 +1,230 @@
+  let ActiveTickets =[], HandeledTickets = [],hiddenCounter1=0 , hiddenCounter2 =0;
+  sortByTime[0].forEach(ticket => {
+    if(ticket.hidden && !ticket.done){
+      hiddenCounter1++;
```
What are those variables? they are not used

Cyber4sPopo commented on client/src/App.js:
```diff
@@ -1,12 +1,230 @@
+        ActiveTickets.push(ticket);
+      }
+  })
+  //const hiddenCounter1 = sortByTime[0].length - (ActiveTickets.length + HandeledTickets.length);
```
Remove

Cyber4sPopo commented on client/src/App.js:
```diff
@@ -1,12 +1,230 @@
+  function revealHidden() {
+    const newArr = sortByTime[0].slice();
+    newArr.forEach((ticket, i) => {
+      if(showFinished){
```
Simplify this `if`/`else`

Cyber4sPopo commented on client/src/App.js:
```diff
@@ -1,12 +1,230 @@
+          </div>
+        </div>
+        {
+          !showFinished
```
Try simplifying this monster

Cyber4sPopo commented on client/src/components/CloseModal.js:
```diff
@@ -0,0 +1,79 @@
+  function handlingTicketClosing() {
+    if (!props.ticketData.done) {
+      axios.post(`/api/tickets/${props.ticketData.id}/done?employe=${employeId}&reason=${closeReason}&additional=${addedInfo}`)
+        .then(() => { props.settingState(false); props.refreshList(props.ticketData.id, true); });
```
Catch errors sent from the server

Cyber4sPopo commented on client/src/components/CloseModal.js:
```diff
@@ -0,0 +1,79 @@
+              </div>
+            )
+          }
+          <div className="employeId" style={{ marginTop: '5%', width: '40%' }}>Additional information (optional)</div>
```
don't use inline style

Cyber4sPopo commented on client/src/components/Ticket.js:
```diff
@@ -0,0 +1,115 @@
+  } = props.ticket;
+  const [closingTicket, setClosingTicket] = useState(false);
+  const smallerContent = content.length>420? content.slice(0,420) : content.slice();
+  const str = content;
```
bad variables names

Cyber4sPopo commented on client/src/components/Ticket.js:
```diff
@@ -0,0 +1,115 @@
+    <div
+      id={`${id}`}
+      className="ticket"
+      style={props.ticket.done ? { backgroundColor: 'rgb(168, 60, 60)' }
```
use class

Cyber4sPopo commented on client/src/components/Ticket.js:
```diff
@@ -0,0 +1,115 @@
+      <div className="ticketContent">
+        {typeof contentToDisplay === 'string' ?contentToDisplay : contentToDisplay.map((line) => <div>{line}</div>)}
+        {
+          // eslint-disable-next-line no-nested-ternary
```
eslint is right. fix this ugly nested ternary.

Cyber4sPopo commented on client/src/myTest.test.js:
```diff
@@ -0,0 +1,80 @@
+    const ActiveListBeforeDone = await page.$eval('#ticketsYouSee', e => e.innerText);
+    console.log(ActiveListBeforeDone);
+
+    await (await page.$('#openTicketModal')).click()
```
nice

Cyber4sPopo reviewed (commented):
please fix

ZBejavu committed:
showing closing and reopening info locally

im making the changes locally in order to not reload th whole list after each handling of a ticket. when i first did it i forgot to changed the properties that hold the information about the ticket handling

ZBejavu committed:
Update README.md

ZBejavu committed:
checking workflow for tests

ZBejavu committed:
Update Checking-tests.yml

ZBejavu committed:
test action

ZBejavu committed:
testing

ZBejavu committed:
Update Checking-tests.yml

ZBejavu committed:
testing idea

ZBejavu committed:
Update Checking-tests.yml

ZBejavu committed:
Update Checking-tests.yml

ZBejavu committed:
Update Checking-tests.yml

ZBejavu committed:
Update Checking-tests.yml

ZBejavu committed:
Update Checking-tests.yml

ZBejavu committed:
Update Checking-tests.yml

ZBejavu committed:
Update Checking-tests.yml

ZBejavu committed:
Update Checking-tests.yml

ZBejavu committed:
Update Checking-tests.yml

ZBejavu committed:
Update Checking-tests.yml

ZBejavu committed:
Update Checking-tests.yml

ZBejavu committed:
Update Checking-tests.yml

ZBejavu committed:
Update Checking-tests.yml

ZBejavu committed:
Update Checking-tests.yml

ZBejavu committed:
Update Checking-tests.yml

ZBejavu committed:
Update Checking-tests.yml

ZBejavu committed:
Update index.css

ZBejavu committed:
Update index.css

ZBejavu committed:
use percy snapshots

ZBejavu committed:
Merge branch 'workBranch' into screenshot-checking

ZBejavu committed:
Update index.css

ZBejavu committed:
Merge branch 'screenshot-checking' of https://github.com/ZBejavu/Final2Wix into screenshot-checking

ZBejavu committed:
Revert "Update index.css"

This reverts commit 70ab3610287d9f694040996403fd66d2dff741e3.

ZBejavu committed:
Merge pull request #8 from ZBejavu/screenshot-checking

Screenshot checking
